x-pack/platform/plugins/shared/dataset_quality/server/rule_types/degraded_docs/executor.ts

```diff
@@ -0,0 +1,179 @@
+import { INDEX, _IGNORED } from '../../../common/es_fields';
+import { generateContext } from '../context';
+import { getDocsStats } from '../get_docs_stats';
+import {
+  ALERT_EVALUATION_CONDITIONS,
+  ALERT_TITLE,
+  DATASET_QUALITY_DATASTREAM_NAME,
+  THRESHOLD_MET_GROUP,
+  type DatasetQualityAlert,
+  type DatasetQualityAlertContext,
+  type DatasetQualityAlertState,
+  type DatasetQualityAllowedActionGroups,
+  type DatasetQualityRuleParams,
+  type DatasetQualityRuleTypeState,
+  AdditionalContext,
+} from '../types';
+
+export const formatDurationFromTimeUnitChar = (time: number, unit: TimeUnitChar): string => {
+  const sForPlural = time !== 0 && time > 1 ? 's' : '';
+  switch (unit) {
+    case 's':
+      return `${time} sec${sForPlural}`;
+    case 'm':
+      return `${time} min${sForPlural}`;
+    case 'h':
+      return `${time} hr${sForPlural}`;
+    case 'd':
+      return `${time} day${sForPlural}`;
+    default:
+      return `${time} ${unit}`;
+  }
+};
+
+export const getRuleExecutor = () =>
+  async function executor(
+    options: RuleExecutorOptions<
+      DatasetQualityRuleParams,
+      DatasetQualityRuleTypeState,
+      DatasetQualityAlertState,
+      DatasetQualityAlertContext,
+      DatasetQualityAllowedActionGroups,
+      DatasetQualityAlert
+    >
+  ): ReturnType<
+    ExecutorType<
+      DatasetQualityRuleParams,
+      DatasetQualityRuleTypeState,
+      DatasetQualityAlertState,
+      DatasetQualityAlertContext,
+      DatasetQualityAllowedActionGroups
+    >
+  > {
+    {
+      const {
+        rule: { name },
+        services,
+        params,
+        logger,
+        getTimeRange,
+      } = options;
+      const { alertsClient, scopedClusterClient } = services;
+
+      if (!alertsClient) {
+        throw new AlertsClientError();
+      }
+
+      const alertLimit = alertsClient.getAlertLimitValue();
+
+      const { dateStart, dateEnd } = getTimeRange(`${params.timeSize}${params.timeUnit}`);
+      const index = params.name;
+
+      const datasetQualityDegradedResults = await getDocsStats({
+        index,
+        dateStart,
+        groupBy: params.groupBy ?? [],
+        query: {
+          must: { exists: { field: _IGNORED } },
+        },
+        scopedClusterClient,
+      });
+
+      const unmetGroupValues: Record<string, string> = {};
+      const compareFn = ComparatorFns.get(params.comparator as Comparator);
+      if (compareFn == null) {
+        throw new Error(
+          i18n.translate('xpack.datasetQuality.rule.invalidComparatorErrorMessage', {
+            defaultMessage: 'invalid thresholdComparator specified: {comparator}',
+            values: {
+              comparator: params.comparator,
+            },
+          })
+        );
+      }
+
+      let generatedAlerts = 0;
+      for (const groupResult of datasetQualityDegradedResults) {
+        const { bucketKey, percentage } = groupResult;
+        const alertId = bucketKey.join('_');
```

Comment:
Another thing to ask Bena about - I believe they are doing a `join()` like this but with `,` and not `-`.

Reply:
Addressed in [4dea2cc](https://github.com/elastic/kibana/pull/216026/commits/4dea2ccd08704628c13a02ab07f56800590eec4d).